aws/resource_aws_msk_cluster_test.go

```diff
@@ -0,0 +1,286 @@
+	availability_zone = "us-east-1b"
+}
+
+resource "aws_subnet" "test_subnet_c" {
+	vpc_id = "${aws_vpc.test_vpc.id}"
+	cidr_block = "10.1.3.0/24"
+	availability_zone = "us-east-1c"
+}
+
+resource "aws_security_group" "test_sg_a" {
+	name        = "allow_all"
+	description = "Allow all inbound traffic"
+	vpc_id      = "${aws_vpc.test_vpc.id}"
+	
+	ingress {
+		from_port   = 0
+		to_port     = 0
+		protocol    = "-1"
+		cidr_blocks = ["0.0.0.0/0"]
+	}
+}  
+
+`, rInt)
+}
+
+func testMskClusterConfig(rInt int) string {
+	return testMskClusterCommonConfig(rInt) + fmt.Sprintf(`
+
+resource "aws_msk_cluster" "test_cluster" {
+	name = "terraform-msk-test-%d"
+	broker_count = 3
+	broker_instance_type = "kafka.m5.large"
+	broker_volume_size = 10
+	kafka_version = "1.1.1"
+	broker_security_groups =["${aws_security_group.test_sg_a.id}"]
+	client_subnets = ["${aws_subnet.test_subnet_a.id}", "${aws_subnet.test_subnet_b.id}", "${aws_subnet.test_subnet_c.id}"]
+}`, rInt)
+}
+
+func testMskClusterConfigWithEncryption(rInt int) string {
+	return testMskClusterCommonConfig(rInt) + fmt.Sprintf(`
+
+resource "aws_kms_key" "test_key" {
+	description             = "KMS key 1"
+	deletion_window_in_days = 10
+}
+  
+resource "aws_msk_cluster" "test_cluster" {
+	name = "terraform-msk-test-%d"
+	broker_count = 3
+	broker_instance_type = "kafka.m5.large"
+	broker_volume_size = 10
+	broker_security_groups =["${aws_security_group.test_sg_a.id}"]
+	kafka_version = "1.1.1"
+	client_subnets = ["${aws_subnet.test_subnet_a.id}", "${aws_subnet.test_subnet_b.id}", "${aws_subnet.test_subnet_c.id}"]
+	encrypt_rest_arn = "${aws_kms_key.test_key.arn}"
+}`, rInt)
+}
+
+func testMskClusterConfigBrokerMonitoring(rInt int, monitoring_type string) string {
+	return testMskClusterCommonConfig(rInt) + fmt.Sprintf(`
+
+resource "aws_msk_cluster" "test_cluster" {
+	name = "terraform-msk-test-%d"
+	broker_count = 3
+	broker_instance_type = "kafka.m5.large"
+	broker_security_groups =["${aws_security_group.test_sg_a.id}"]
+	broker_volume_size = 10
+	kafka_version = "1.1.1"
+	client_subnets = ["${aws_subnet.test_subnet_a.id}", "${aws_subnet.test_subnet_b.id}", "${aws_subnet.test_subnet_c.id}"]
+	enhanced_monitoring = "%s"
+}`, rInt, monitoring_type)
+}
+
+func testAccCheckMskClusterExists(n string, cluster *kafka.ClusterInfo) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		rs, ok := s.RootModule().Resources[n]
+		if !ok {
+			return fmt.Errorf("Not found: %s", n)
+		}
+
+		if rs.Primary.ID == "" {
+			return fmt.Errorf("No MSK Cluster ID is set")
+		}
+
+		conn := testAccProvider.Meta().(*AWSClient).kafkaconn
+		describeOpts := &kafka.DescribeClusterInput{
+			ClusterArn: aws.String(rs.Primary.Attributes["arn"]),
+		}
+		resp, err := conn.DescribeCluster(describeOpts)
+		if err != nil {
+			return err
+		}
+
+		*cluster = *resp.ClusterInfo
+
+		return nil
+	}
+}
+
+func testAccCheckAWSMskClusterAttributes(cluster *kafka.ClusterInfo) resource.TestCheckFunc {
```

Comment:
Combined with the `ImportStateVerify: true` testing, which will automatically check all state attributes values against a fresh API read, we generally prefer to check the resource state values via [`resource.TestCheckResourceAttr()`](https://godoc.org/github.com/hashicorp/terraform/helper/resource#TestCheckResourceAttr) instead of implementing custom test check functions since they tend to become out of sync easily. e.g.

```go
resource.TestCheckResourceAttr("aws_msk_cluster.test_cluster", "broker_count", "3"),
resource.TestCheckResourceAttr("aws_msk_cluster.test_cluster", "broker_instance_type", "kafka.m5.large"),
resource.TestCheckResourceAttr("aws_msk_cluster.test_cluster", "broker_volume_size", "10"),
resource.TestCheckResourceAttr("aws_msk_cluster.test_cluster", "kafka_version", "1.1.1"),
```

To check the ARN:

```go
testAccMatchResourceAttrRegionalARN("aws_msk_cluster", "arn", "kafka", regexp.MustCompile(`cluster/.+`)),
```

To check the value matches another resource attribute:

```go
resource.TestCheckResourceAttrPair("aws_msk_cluster.test_cluster", "encrypt_rest_arn", "aws_kms_key.test_key", "arn"),
```

Reply:
@dthtvwls if you can push up what you have so far, I can take a look.

`TestCheckResourceAttr` and friends automatically use the available Terraform state applied by the `TestStep` configuration. Here is a full example of the acceptance testing with these in action:

https://github.com/terraform-providers/terraform-provider-aws/blob/409a3dc42d1828339cd3ceaff3317082f3d7ba04/aws/resource_aws_ram_resource_share_test.go#L16-L43